airflow/ti_deps/deps/trigger_rule_dep.py

```diff
@@ -0,0 +1,179 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from sqlalchemy import case, func
+
+import airflow
+from airflow.ti_deps.deps.base_ti_dep import BaseTIDep
+from airflow.utils.db import provide_session
+from airflow.utils.state import State
+
+
+class TriggerRuleDep(BaseTIDep):
+    """
+    Determines if a task's upstream tasks are in a state that allows a given task instance
+    to run.
+    """
+    NAME = "Trigger Rule"
+    IGNOREABLE = True
+    IS_TASK_DEP = True
+
+    @provide_session
+    def _get_dep_statuses(self, ti, session, dep_context):
+        TI = airflow.models.TaskInstance
+        TR = airflow.models.TriggerRule
+
+        # Checking that all upstream dependencies have succeeded
+        if not ti.task.upstream_list:
+            yield self._passing_status(
+                reason="The task instance did not have any upstream tasks.")
+            raise StopIteration
+
+        if ti.task.trigger_rule == TR.DUMMY:
+            yield self._passing_status(reason="The task had a dummy trigger rule set.")
+            raise StopIteration
+
+        # TODO(unknown): this query becomes quite expensive with dags that have many
```

Comment:
That would be my comment. The aggregation query is very expensive and ideally we should get rid of it. It takes around 0.1s per evaluated task. It is one of the main reasons the scheduler will get slower with a large amount of tasks that need to be considered. 

Sent from my iPhone

> On 26 aug. 2016, at 22:20, aoen <notifications@github.com> wrote:
> 
> In airflow/ti_deps/deps/trigger_rule_dep.py:
> 
> > +    @provide_session
> > +    def _get_dep_statuses(self, ti, session, dep_context):
> > +        TI = airflow.models.TaskInstance
> > +        TR = airflow.models.TriggerRule
> > +
> > +        # Checking that all upstream dependencies have succeeded
> > +        if not ti.task.upstream_list:
> > +            yield self._passing_status(
> > +                reason="The task instance did not have any upstream tasks.")
> > +            raise StopIteration
> > +
> > +        if ti.task.trigger_rule == TR.DUMMY:
> > +            yield self._passing_status(reason="The task had a dummy trigger rule set.")
> > +            raise StopIteration
> > +
> > +        # TODO(unknown): this query becomes quite expensive with dags that have many
> I don't know, I didn't add this comment (just annotated with (unknown))
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>